src/main/java/com/gluonhq/richtext/RichTextAreaSkin.java

```diff
@@ -156,19 +161,49 @@ public void dispose() {
 
     /// PRIVATE METHODS /////////////////////////////////////////////////////////
 
-    //TODO Need more optimal way of rendering text fragments.
+    // TODO Need more optimal way of rendering text fragments.
     //  For now rebuilding the whole text flow
     private void refreshTextFlow() {
         fontCacheEvictionTimer.pause();
         try {
             var fragments = new ArrayList<Text>();
-            viewModel.walkFragments((text, decoration) -> fragments.add(buildText(text, decoration)));
+            var backgroundIndexRanges = new ArrayList<IndexRangeColor>();
+            var length = new AtomicInteger();
+            viewModel.walkFragments((text, decoration) -> {
+                final Text textNode = buildText(text, decoration);
+                fragments.add(textNode);
+
+                if (decoration.getBackground() != Color.TRANSPARENT) {
+                    final IndexRangeColor indexRangeColor = new IndexRangeColor(
+                            length.get(),
+                            length.get() + textNode.getText().length(),
+                            decoration.getBackground()
+                    );
+                    backgroundIndexRanges.add(indexRangeColor);
+                }
+                length.addAndGet(textNode.getText().length());
+            });
             textFlow.getChildren().setAll(fragments);
+            addBackgroundPathsToLayers(backgroundIndexRanges);
         } finally {
             fontCacheEvictionTimer.start();
         }
     }
 
+    private void addBackgroundPathsToLayers(List<IndexRangeColor> backgroundIndexRanges) {
+        if (layers.getChildren().size() > 3) {
+            layers.getChildren().removeIf(node -> node.getStyleClass().contains("background"));
+        }
+        backgroundIndexRanges.forEach(indexRangeBackground -> {
+            final Path path = new Path(textFlow.rangeShape(indexRangeBackground.getStart(), indexRangeBackground.getEnd()));
+            path.setStrokeWidth(0);
+            path.setFill(indexRangeBackground.getColor());
+            path.getStyleClass().add("background");
```

Comment:
why a style class? Is it required for css styling as part of the control internals?

If it is only for identification, better use `setId`? That should be more performant